lib/tbot/spiffe/trust_bundle_cache.go

```diff
@@ -0,0 +1,614 @@
+			// We'll simply ignore this since it would put consumers into a
+			// weird state to not have a trust bundle for the local trust domain
+			// available.
+			log.WarnContext(
+				ctx,
+				"Ignoring event indicating CA deletion",
+			)
+		case types.KindSPIFFEFederation:
+			_, ok := bundleSet.Federated[event.Resource.GetName()]
+			if !ok {
+				// Doesn't exist locally so nothing to change...
+				return
+			}
+			log.InfoContext(
+				ctx, "Processed deletion for federated trust bundle",
+			)
+			delete(bundleSet.Federated, event.Resource.GetName())
+			m.setAndBroadcastBundleSet(bundleSet)
+		}
+	case types.OpPut:
+		switch event.Resource.GetKind() {
+		case types.KindCertAuthority:
+			ca, ok := event.Resource.(types.CertAuthority)
+			if !ok {
+				log.WarnContext(
+					ctx,
+					"Event did not contain expected resource type",
+					"got", reflect.TypeOf(event.Resource),
+				)
+				return
+			}
+			if ca.GetType() != types.SPIFFECA {
+				// Safeguard against receiving an event not for the SPIFFE CA.
+				log.WarnContext(
+					ctx,
+					"Ignoring event for non-SPIFFE CA",
+					"type", ca.GetType(),
+				)
+				return
+			}
+			if ca.GetClusterName() != m.clusterName {
+				// Safeguard against receiving an event for a different cluster.
+				log.WarnContext(
+					ctx,
+					"Ignoring event for different cluster",
+					"cluster", ca.GetClusterName(),
+				)
+				return
+			}
+
+			bundle, err := convertSPIFFECAToBundle(ca)
+			if err != nil {
+				// This is "bad". Ideally, this situation should never occur,
+				// but if it does, it's preferable that subscribed workloads
+				// continue to use the last good bundle.
+				log.WarnContext(
+					ctx,
+					"Failed to convert SPIFFE CA to trust bundle",
+					"error", err,
+				)
+				return
+			}
+
+			if bundleSet.Local.Equal(bundle) {
+				log.DebugContext(
+					ctx,
+					"Event resulted in no change to local trust bundle, ignoring",
+				)
+				return
+			}
+			log.InfoContext(ctx, "Processed update for local trust bundle")
+			bundleSet.Local = bundle
+			m.setAndBroadcastBundleSet(bundleSet)
+		case types.KindSPIFFEFederation:
+			r153, ok := event.Resource.(types.Resource153Unwrapper)
+			if !ok {
+				log.WarnContext(
+					ctx,
+					"Event did not contain a 153 style resource",
+					"got", reflect.TypeOf(event.Resource),
+				)
+				return
+			}
+			federation, ok := r153.Unwrap().(*machineidv1pb.SPIFFEFederation)
+			if !ok {
+				log.WarnContext(
+					ctx,
+					"Event did not contain expected type",
+					"got", reflect.TypeOf(event.Resource),
+				)
+				return
+			}
+			bundle, err := convertSPIFFEFederationToBundle(federation)
+			if err != nil {
+				// TODO: Should we match the behavior for the local trust
+				// bundle that's derived from the CA - i.e continue to use the
+				// last good bundle, or, should we remove this from our local
+				// set and tell workloads to start ignoring this trust domain?
```

Comment:
Hmm, I can see an argument for both. If something's wrong with the bundle, I assume it will remain bad until at least the next server-side sync? (Which is itself determined by the refresh hint?)

From the user's PoV I imagine keeping the last good bundle should keep things working for a bit longer, and if we removed it we'd mostly be swapping one error for another. I'm not sure what other advantages removal would have?

Reply:
Yeah - realistically, it shouldn't even get into this state because the bundle is validated by the auth server before persisting it. It mostly just felt like actively removing the bundle was drastic - that decision will immediately be pushed out to all subscribers.

My general design plan with most of this stuff has been to try and have it go into degraded/recovering states rather than failure states. We can likely build on this in the future a bit to detect an extended "degraded" state and then take further action.

There's some work we probably need to do on the rest of tbot to try and refocus on going into a degrading/recovering states rather than crashing out. Might be nice to make some things (e.g handling the renewal of the bot identity etc) a bit more of a state machine in some regards.